hw/ip/edn/doc/_index.md

```diff
@@ -0,0 +1,180 @@
+---
+title: "EDN HWIP Technical Specification"
+---
+
+# Overview
+
+This document specifies EDN hardware IP functionality.
+This module conforms to the [Comportable guideline for peripheral functionality.]({{< relref "doc/rm/comportability_specification" >}})
+
+
+## Features
+
+The Entropy Distribution Network (EDN) block provides both hardware and software interfaces to the [CSRNG IP]({{< relref "hw/ip/csrng/doc" >}}) module.
+A primary objective of the EDN block is to provide a simpler hardware interface for the peripheral blocks to use, in which case they do not have to directly interface with the CSRNG module.
+- The EDN block provides a set of registers for firmware to manage a CSRNG application port.
+- There are four request/acknowledge hardware interfaces, and one can be configured to act as a broadcast type of interface (only 1 of 4 interfaces).
```

Comment:
Maybe I'm not thinking creatively enough, but what's the use case for the broadcast mode?  It seems counter-intuitive to share any entropy / use it more than once.

Reply:
I agree.  However at this point, I'm more inclined to put it as an option in the spec and argue about it at top level.

Reply:
Would it make sense to have some sort of ID scheme? So you have (say) 4 devices hanging off the broadcast port and every chunk of entropy that gets broadcast has an ID attached (0 to 3) to tell the devices whether they should use it. Then we can avoid more than one device using the same entropy. We could also make the weighting between devices configurable: you could do something super-simple like have a 32-bit register containing 16 2-bit values to be used as successive device IDs. I doubt we'd ever want anything more configurable than that.

Reply:
Isn't that awfully close to just having a dedicated entropy interface, but sharing some wires and having a more complicated protocol than dedicated 1:1 entropy distribution would entail?

(Martin, I'm OK to leave it as an EDN option, I just want to make sure I understand what we specifically are doing with the broadcast mode in the design.)

Reply:
TBH, I agree with you, Chris, it is could be troublesome.  As far as I can tell it started as a brainstorming idea, from an IP design (maybe AES?).  However @mwbranstad has had exchanges  people on this, and it seems to have legs.  Though Mark may have to weigh in here in 1.5 weeks.

And I think I agree that we don't want to add complexity to this feature making it more complicated than the standard interface.

The real advantage seems to be that it offers timed playback: consumers get entropy at some predetermined interval.    It seems like a simple way of throttling entropy consumption.  Broadcast consumers want a continuous drip of entropy, and they presumably for whatever reason don't need it to be secret (a claim worth inspecting!).    So then, some top-level management utility works out how much entropy can go to the "Broadcast consumers" vs. the presumably more critical consumers who have burst entropy consumption requirements.  The timer is adjusted by a central firmware utility to regulate this channel.

In the end it seems a bit chaotic as a fairness mechanism, but maybe fairness mechanisms are all chaotic.   Also, no matter how we manage this starvation is going to be a *huge* problem no matter what we do.   All of these obfuscation mechanisms are going to want to consume far more entropy than one little non-pipelined AES can provide.

Reply:
Maybe we can leave it in as a "proposed" feature? and then open yet another issue to track these top level consumption issues (or add it as a thread an existing issue)?

Reply:
i should clarify the broadcast option was just meant for blocks that locally have an LFSR, and just care to have SOME randomness and not necessarily very "good" randomness.

Does that make sense? You could make these all 1:1 connections if we choose, but i think a fair number of sub blocks do not require entropy good enough for masking.

Reply:
Edit: "- For consumers which desire entropy for lower-secrecy tasks (i.e. LFSR reseeeding), one of the four hardware interfaces can be configured to act as a broadcast interface to serve muliple consumers from a shared random number stream"